pkg/reconciler/certificate/certificate.go

```diff
@@ -116,6 +121,17 @@ func (c *Reconciler) reconcile(ctx context.Context, knCert *v1alpha1.Certificate
 		knCert.Status.MarkNotReady(cmCertReadyCondition.Reason, cmCertReadyCondition.Message)
 		return c.setHTTP01Challenges(knCert, cmCert)
 	case cmCertReadyCondition.Status == cmmeta.ConditionTrue:
+		if cmCertRenewingCondition != nil && cmCertRenewingCondition.Reason == renewingEvent {
```

Comment:
So the `renewingEvent` (`Renewing`) is from https://github.com/cert-manager/cert-manager/blob/master/internal/controller/certificates/policies/constants.go#L41-L43 but we cannot import the internal lib so we defined it in Line#59?
And if cert-manager changed the `Renewing` word, our code will break and nobody notice it, won't it.

So, I still think that we should trigger the reconile loop by `NotAfter` instead of the `Renewing` as https://github.com/knative-sandbox/net-certmanager/pull/417/files#r901272781 commented.

Reply:
@nak3 
> So the renewingEvent (Renewing) is from https://github.com/cert-manager/cert-manager/blob/master/internal/controller/certificates/policies/constants.go#L41-L43 but we cannot import the internal lib so we defined it in Line#59?

yes

> And if cert-manager changed the Renewing word, our code will break and nobody notice it, won't it.

I don't think the code will break, but definitely will not notice it, I think it will have the current behaviour

> So, I still think that we should trigger the reconile loop by NotAfter instead of the Renewing as https://github.com/knative-sandbox/net-certmanager/pull/417/files#r901272781 commented.

`NotAfter` field indicates when the certificate will expire, not when it should be renewing. I'm not sure if I fully understand your comment, but you would like something like this, right?
```
knCert.Status.NotAfter = cmCert.Status.RenewalTime
```
and check the diff between `knCert.Status.NotAfter` and the current time

Reply:
> NotAfter field indicates when the certificate will expire, not when it should be renewing.

I assumed that cert-manager updates the `NotAfter` field when it renewed the certificate. If the assumption is true, we can add the trigger to update the status like this?

```go
	// Line around 106
	renewed := false
	if knCert.Status.NotAfter != cmCert.Status.NotAfter {
		renwed = true
	}
	knCert.Status.NotAfter = cmCert.Status.NotAfter

	//  ... snip ... //

	switch {
	case cmCertReadyCondition.Status == cmmeta.ConditionTrue:
		if renewed {
			// set renew condition
			renewCondition := apis.Condition{
				Type:   renewingEvent,
				Status: corev1.ConditionTrue,
			}
			certificateCondSet.Manage(&knCert.Status).SetCondition(renewCondition)
			return c.setHTTP01Challenges(knCert, cmCert)
		}
		// ...
	}
```

Reply:
> I assumed that cert-manager updates the NotAfter field when it renewed the certificate

Yes, but the cert-manager certificate will never be renewed if we do not tell the ingress to setup HTTP challenges routes. Therefore, cert-manager `NotAfter` field won't change until the certificate expires, meaning `knCert.Status.NotAfter != cmCert.Status.NotAfter` will never happen before expiration

Maybe something like that will work (I'm testing it)
```
        // Propagate cert-manager Certificate status to Knative Certificate.
	renewalTime := cmCert.Status.RenewalTime
	now := time.Now()
	if now.Before(renewalTime.Time) {
		knCert.Status.NotAfter = renewalTime
	} else {
		knCert.Status.NotAfter = nil
	}
```